src/ripple/basics/promises.h

```diff
@@ -0,0 +1,958 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2023 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_BASICS_PROMISES_H_INCLUDED
+#define RIPPLE_BASICS_PROMISES_H_INCLUDED
+
+#include <atomic>
+#include <cassert>
+#include <exception>
+#include <functional>
+#include <memory>
+#include <stdexcept>
+#include <string>
+#include <tuple>
+#include <vector>
+
+namespace ripple {
+
+template <std::size_t I, typename... Ts>
+struct nth_type : public std::tuple_element<I, std::tuple<Ts...>>
+{
+};
+
+template <std::size_t I, typename... Ts>
+using nth_type_t = typename nth_type<I, Ts...>::type;
+
+namespace detail {
+
+struct construct_callbacks
+{
+};
+struct construct_value
+{
+};
+struct construct_error
+{
+};
+
+template <typename T>
+struct Value
+{
+    T value_;
+
+    template <typename... Args>
+    Value(Args&&... args) : value_(std::forward<Args>(args)...)
+    {
+    }
+
+    T const&
+    get() const
+    {
+        return value_;
+    }
+
+    T&
+    get()
+    {
+        return value_;
+    }
+};
+
+template <>
+struct Value<void>
+{
+    void
+    get() const
+    {
+    }
+};
+
+template <typename T, typename C, typename V>
+union Storage
```

Comment:
`std::variant` cannot hold `void`. `Value` here is specialized for `void`.

Reply:
The `variant` wouldn't hold `void`, it would hold `Value<void>`.  So I guess I'm not completely understanding.  Or perhaps my question was ambiguous.  I'm thinking about something along the lines of:
```c++
template <typename T, typename C, typename V>
class Storage
{
public:
    using callback_type = C;
    using value_type = V;
    using error_type = std::exception_ptr;
private:
   std::variant<
         std::vector<callback_type>
       , std::shared_ptr<T>
       , Value<value_type>
       , error_type
   > data_;
...
```

Or if the API of `std::variant` was sufficient then maybe expose the `variant` directly instead of wrap it in a class:

```c++
template <typename T, typename C, typename V>
using Storage = std::variant<
                               std::vector<C>
                             , std::shared_ptr<T>
                             , Value<V>
                             , std::exception_ptr
                            >;
```

Disclaimer:  I haven't prototyped either of these suggestions.  I just know that `union`s  are tricky to work with directly, and that was the motivation for introducing `variant` into the `std::lib`.